### DISCUSSION TITLE
TPPL Installation

### DISCUSSION THREAD

**Discussion by vsenderov** [treeppl/treeppl#74] [Announcements]
Thanks @Anders Ågren Thuné for today's discussion. I think we have a fairly good understanding of how Miking packaging works right now. To summarize, we have a way of supplying portable versions of Miking on Linux and Mac using the Nix package manager (only a prerequisite for the packager, but not for the user).  One way to make use of this functionality (not the only way), would be to write a modified version of it that supplies not just Miking, but also Miking DPPL, and TPPL.  This means writing Nix descriptions for Miking DPPL and TPPL, and changing the packaging script to pack all three instead of just Miking. Also this extended script should probably be separate from Miking as it is includes non-Miking components.  Finally, the Python package should use the portable version of TPPL that we just packaged.  The challenge here is that the Python package should know the location of the TPPL portable, which could either be a default (like /opt/treeppl), or there could be some user-interaction during installation.  In any case, the user will need to

Unpack the TPPL tarball
Install the Python package
Indicate to the Python package where the TPPL tarball is

There are drawbacks:

the size pretty is pretty large
the Python package will need to be manually updated if the tarball is re-installed

But I don't think any of these are deal breakers.  Size is not so much of an issue nowadays.  And reinstalling the Python package after upgrading TPPL is logical and we can inform the users about it. Furthermore, most likely the Python package will not have to be reinstalled for small changes in TPPL, except when the I/O is changed, which should happen very rarely.  In the future, when we have .deb, brew, etc. we can replace the portable version with globally installed versions.

Anders and David should probably brainstorm this a little further, and if you believe this is the way to go than we should plan on implementing this.

PS: We don't worry about Windows, as WSL + Vscode = Linux

**Comment by elegios** [comment]
In case it is relevant, because I don't think I've pushed this anywhere, I have a modified version of the `miking` nix package that includes functionality to automatically populate the `MCORE_LIBS` variable appropriately to discover libraries (`setupHook` in particular), as well as a specification for `miking-dppl`. Here it is in the nix flake I use when I work on treeppl:

```nix
{
  description = "TreePPL";
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
    flake-utils.url = "github:numtide/flake-utils";
    miking-src = {
      url = "github:miking-lang/miking";
      flake = false;
    };
    miking-dppl-src = {
      url = "github:miking-lang/miking-dppl";
      flake = false;
    };
  };
  outputs = { self, nixpkgs, flake-utils, miking-src, miking-dppl-src, ... }@inputs:
    with flake-utils.lib; eachSystem allSystems (system:
      let
        pkgs = nixpkgs.legacyPackages.${system};
        ocamlPackages = pkgs.ocaml-ng.ocamlPackages_5_0;
        miking = pkgs.stdenv.mkDerivation (finalAttrs: rec {
          pname = "miking";
          version = "0.0.0+git";

          withLwt = true;
          withToml = true;
          withOwl = true;

          src = miking-src;

          propagatedBuildInputs = [
            pkgs.coreutils
            ocamlPackages.ocaml
            ocamlPackages.findlib
            ocamlPackages.dune_3
            ocamlPackages.linenoise
            ocamlPackages.menhir
          ] ++ pkgs.lib.lists.optional finalAttrs.withLwt ocamlPackages.lwt
          ++ pkgs.lib.lists.optional finalAttrs.withOwl ocamlPackages.owl
          ++ pkgs.lib.lists.optional finalAttrs.withToml ocamlPackages.toml;
          nativeBuildInputs = [
            pkgs.makeWrapper
            pkgs.time
          ];

          makeFlags = [ "prefix=$(out)" "ocamllibdir=$(out)/lib/ocaml/${ocamlPackages.ocaml.version}/site-lib" ];

          postInstall = ''
            wrapProgram $out/bin/mi \
              --suffix PATH : ${pkgs.coreutils}/bin
          '';

          setupHook = pkgs.writeText "setupHook.sh" ''
            addMCorePath() {
              echo test $1
              for dir in "''$1"/lib/mcore/*; do
                export MCORE_LIBS="''${MCORE_LIBS-}''${MCORE_LIBS:+:}''$(basename ''$dir)=''$dir"
              done
            }
            addEnvHooks "$targetOffset" addMCorePath
          '';
        });
        miking-dppl = pkgs.stdenv.mkDerivation rec {
          pname = "miking-dppl";
          version = "0.0.0+git";

          src = miking-dppl-src;

          propagatedBuildInputs = [ miking ];

          makeFlags = [ "BIN_PATH=$(out)/bin/" "SRC_PATH=$(out)/lib/mcore/coreppl" ];

          installTargets = [ "install-coreppl" ];
        };
      in rec {
        packages.default = pkgs.stdenv.mkDerivation rec {
          pname = "treeppl";
          version = "0.0.0+git";

          src = ./.;

          propagatedBuildInputs = [ miking-dppl ];
        };
        devShells.default = pkgs.mkShell {
          name = "TreePPL dev shell";

          inputsFrom = [packages.default];
          buildInputs = [pkgs.gdb];

          env = {
            MIDPPL_SRC = "${miking-dppl}/lib/mcore/coreppl";
          };
        };
      });
    }
```

I should probably push this at some point, but I haven't taken the time to figure out what the nicest reusable way is, for now it's just been convenient to have everything in one file for each project.

**Comment by vsenderov** [comment]
Based on https://github.com/treeppl/treeppl/wiki/TPPL-Minutes-from-2024-08-27, I am adding some notes here

- Anders has packaged Miking and OCaml as a stand-alone using the NIX package manager
- Miking-DPPL will be part of Miking
- There will be a release system for Miking
    - Miking + Miking-DPPL will have a release candidate
    - As soon as this is available, we will test whether TreePPL works with the release candidate
    - If yes, then we will make an official release of Miking and update the releases of TreePPL-Python and TreePPL-R to require this specific Miking version
- TreePPL release system
    - We have the core TreePPL compiler in its own repo, this does not get an official release
    - Have Python TreePPL and R TreePPL, which package the TreePPL compiler as data and those get released
- Perhaps getting rid of the environment variables (only `mi` and `tpplc` should be dependencies)?

**Comment by ThimotheeV** [comment]
I have proposal about the current installation [guide](https://github.com/treeppl/treeppl?tab=readme-ov-file#installation) :
- I would like to add a full and explicit dependencies list because for now I don't think that the case (whith maybe a link or a distinct section for installation).
- Maybe a command line example without python or r package will be interesting to have. 
- Can we move all the explanation on the TreePPL web site and just have a link in the readme, that will be much more simple to maintain ?

**Reply by vsenderov** [reply]
I am wondering if the guide you pointed to is simply outdated? The intended installations guides by me and @vresbok are currently already uploaded to the website under https://treeppl.org/docs/Howtos/linux. I guess it does not hurt to sync them with the one in the source code you mentioned and extend them further with your suggestions if you prefer.
